src/GaussianRenderer.cpp

```diff
@@ -0,0 +1,317 @@
+
+typedef Eigen::Matrix<float, 3, 1, Eigen::DontAlign> Vector3f;
+typedef Eigen::Matrix<int, 3, 1, Eigen::DontAlign>   Vector3i;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+float sigmoid(const float m1) {
+  return 1.0f / (1.0f + exp(-m1));
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#define CUDA_SAFE_CALL_ALWAYS(A)                                                                   \
+  A;                                                                                               \
+  cudaDeviceSynchronize();                                                                         \
+  if (cudaPeekAtLastError() != cudaSuccess)                                                        \
+    logger().error(cudaGetErrorString(cudaGetLastError()));
+
+#if DEBUG || _DEBUG
+#define CUDA_SAFE_CALL(A) CUDA_SAFE_CALL_ALWAYS(A)
+#else
+#define CUDA_SAFE_CALL(A) A
+#endif
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// Load the Gaussians from the given file. This comes more or less unmodified from here:
+// https://gitlab.inria.fr/sibr/sibr_core/-/blob/fossa_compatibility/src/projects/gaussianviewer/renderer/GaussianView.cpp
+template <int D>
+int loadPly(const char* filename, std::vector<GaussianData::Pos>& pos,
+    std::vector<GaussianData::SHs<3>>& shs, std::vector<float>& opacities,
+    std::vector<GaussianData::Scale>& scales, std::vector<GaussianData::Rot>& rot, Vector3f& minn,
+    Vector3f& maxx) {
+  std::ifstream infile(filename, std::ios_base::binary);
+
+  if (!infile.good()) {
+    logger().error("Unable to find model's PLY file, attempted: {}", filename);
+  }
+
+  // "Parse" header (it has to be a specific format anyway).
+  std::string buff;
+  std::getline(infile, buff);
+  std::getline(infile, buff);
+
+  std::string dummy;
+  std::getline(infile, buff);
+  std::stringstream ss(buff);
+  int               count;
+  ss >> dummy >> dummy >> count;
+
+  // Output number of Gaussians contained.
+  logger().info("Loading {} Gaussian splats", count);
+
+  while (std::getline(infile, buff)) {
+    if (buff.compare("end_header") == 0) {
+      break;
+    }
+  }
+
+  // Read all Gaussians at once (AoS).
+  std::vector<GaussianData::RichPoint<D>> points(count);
+  infile.read((char*)points.data(), count * sizeof(GaussianData::RichPoint<D>));
+
+  // Resize our SoA data.
+  pos.resize(count);
+  shs.resize(count);
+  scales.resize(count);
+  rot.resize(count);
+  opacities.resize(count);
+
+  // Gaussians are done training, they won't move anymore. Arrange
+  // them according to 3D Morton order. This means better cache
+  // behavior for reading Gaussians that end up in the same tile
+  // (close in 3D --> close in 2D).
+  minn = Vector3f(FLT_MAX, FLT_MAX, FLT_MAX);
```

Comment:
We shouldn't use the C standard library.

```suggestion
  minn = Vector3f(std::numeric_limits<float>::max(), std::numeric_limits<float>::max(),
      std::numeric_limits<float>::max());
```